server/src/main/java/org/opensearch/index/store/DirectoryFileTransferTracker.java

```diff
@@ -0,0 +1,187 @@
+
+    private volatile long lastSuccessfulSegmentDownloadBytes;
+
+    /**
+     * Provides moving average over the last N upload speed (in bytes/s) of segment files downloaded from the source {@link org.apache.lucene.store.Directory}.
+     * N is window size
+     */
+    private volatile MovingAverage downloadBytesPerSecMovingAverageReference;
+
+    private final int DIRECTORY_FILES_TRANSFER_DEFAULT_WINDOW_SIZE = 20;
+
+    public long getDownloadBytesStarted() {
+        return downloadBytesStarted;
+    }
+
+    public void addDownloadBytesStarted(long size) {
+        downloadBytesStarted += size;
+    }
+
+    public long getDownloadBytesFailed() {
+        return downloadBytesFailed;
+    }
+
+    public void addDownloadBytesFailed(long size) {
+        downloadBytesFailed += size;
+    }
+
+    public long getDownloadBytesSucceeded() {
+        return downloadBytesSucceeded;
+    }
+
+    public void addDownloadBytesSucceeded(long size) {
+        downloadBytesSucceeded += size;
+    }
+
+    public boolean isDownloadBytesPerSecAverageReady() {
+        return downloadBytesPerSecMovingAverageReference.isReady();
+    }
+
+    public double getDownloadBytesPerSecAverage() {
+        return downloadBytesPerSecMovingAverageReference.getAverage();
+    }
+
+    public void addDownloadBytesPerSec(long bytesPerSec) {
+        this.downloadBytesPerSecMovingAverageReference.record(bytesPerSec);
+    }
+
+    public boolean isDownloadBytesAverageReady() {
+        return downloadBytesMovingAverageReference.isReady();
+    }
+
+    public double getDownloadBytesAverage() {
+        return downloadBytesMovingAverageReference.getAverage();
+    }
+
+    public void updateLastDownloadedSegmentSize(long size) {
+        lastSuccessfulSegmentDownloadBytes = size;
+        this.downloadBytesMovingAverageReference.record(size);
+    }
+
+    public long getLastDownloadTimestampMs() {
+        return lastDownloadTimestampMs;
+    }
+
+    public void updateLastDownloadTimestampMs(long downloadTimestampInMs) {
+        this.lastDownloadTimestampMs = downloadTimestampInMs;
+    }
+
+    public DirectoryFileTransferTracker() {
+        downloadBytesMovingAverageReference = new MovingAverage(DIRECTORY_FILES_TRANSFER_DEFAULT_WINDOW_SIZE);
+        downloadBytesPerSecMovingAverageReference = new MovingAverage(DIRECTORY_FILES_TRANSFER_DEFAULT_WINDOW_SIZE);
+    }
+
+    public DirectoryFileTransferTracker.Stats stats() {
+        return new Stats(
+            downloadBytesStarted,
+            downloadBytesFailed,
+            downloadBytesSucceeded,
+            lastDownloadTimestampMs,
+            downloadBytesMovingAverageReference.getAverage(),
+            lastSuccessfulSegmentDownloadBytes,
+            downloadBytesPerSecMovingAverageReference.getAverage()
+        );
+    }
+
+    /**
+     * Represents the tracker's stats presentable to an API.
+     *
+     * @opensearch.internal
+     */
+    public static class Stats implements Writeable {
```

Comment:
Not sure what purpose `Stats` is solving if it has exact set of fields as tracker? Not saying we shouldn't have it, trying to understand its purpose.

Reply:
The separate `Stats` inner class is to ensure SerDe capabilities from an XContent since these fields are being used in an API output directly